src/inmanta/ast/statements/generator.py

```diff
@@ -360,26 +664,33 @@ def get_all_eager_promises(self) -> Iterator["StaticEagerPromise"]:
     def requires(self) -> List[str]:
         return list(chain.from_iterable(sub.requires() for sub in [self.condition, self.if_expression, self.else_expression]))
 
-    def requires_emit(self, resolver: Resolver, queue: QueueScheduler) -> Dict[object, VariableABC]:
-        requires: Dict[object, VariableABC] = super().requires_emit(resolver, queue)
+    def requires_emit(
+        self, resolver: Resolver, queue: QueueScheduler, *, lhs: Optional[ResultCollector[object]] = None
+    ) -> dict[object, VariableABC[object]]:
+        requires: Dict[object, VariableABC[object]] = super().requires_emit(resolver, queue)
 
         # This ResultVariable will receive the result of this expression
-        result: ResultVariable = ResultVariable()
+        result: ResultVariable[object] = ResultVariable()
 
         # Schedule execution to resume when the condition can be executed
-        resumer: RawResumer = ConditionalExpressionResumer(self, result)
+        resumer: RawResumer = ConditionalExpressionResumer(self, result, lhs=lhs)
         self.copy_location(resumer)
         RawUnit(queue, resolver, self.condition.requires_emit(resolver, queue), resumer)
 
         # Wait for the result variable to be populated
         requires[self] = result
         return requires
 
+    def requires_emit_gradual(
+        self, resolver: Resolver, queue: QueueScheduler, resultcollector: ResultCollector[object]
+    ) -> dict[object, VariableABC[object]]:
+        return self.requires_emit(resolver, queue, lhs=resultcollector)
```

Comment:
I made the conditional expression gradually executable while I was at it.

Reply:
what does that even mean? conditional expressions produce only one value, so why is it a benefit to be gradual?

Reply:
They resolve to one expression, which may produce more than one value. For example, the conditional expression `x.myrelation = condition ? y.a : y.b` obviously has to wait for the condition to fully execute (we can't be gradual there), but once it has picked a branch it can pass on the result collector to the relevant child expression. With the implementation on master this expression would block the gradual execution of `y.a` or `y.b`. The test case `test_conditional_expression_gradual` I added illustrates this: it fails on master but succeeds on this branch.